### PR TITLE
Ignore global phase when canonicalizing EigenGate's circuit diagram exponent

### PR DESCRIPTION
**Pull request by Strilanc**
Fixes https://github.com/quantumlib/Cirq/issues/1080

### CODE REVIEW

File: cirq/ops/eigen_gate.py [hunk 1]
```diff
@@ -105,6 +105,63 @@ def _with_exponent(self: TSelf,
...
+
+        return result
+
+    # virtual method
+    def _eigen_shifts(self) -> List[float]:
+        """Describes the eigenvalues of the gate's matrix.
+
+        By default, this just extracts the shifts by calling
+        self._eigen_components(). However, because that method generates
+        matrices it may be extremely expensive.
```
[the remark this answers is not in the document]
Comment (dabacon) [reply cirq/ops/eigen_gate.py]: Maybe there should be some testing methods that you call on all eigengates to ensure that they don't have this problem?

[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/eigen_gate.py]: Added this to https://github.com/quantumlib/Cirq/issues/1053 (the standard gate test battery)

File: cirq/ops/eigen_gate.py [hunk 2]
```diff
@@ -105,6 +105,63 @@ def _with_exponent(self: TSelf,
...
         # pylint: enable=unexpected-keyword-arg
 
+    def _diagram_exponent(self, args: protocols.CircuitDiagramInfoArgs):
+        """The exponent to use in circuit diagrams.
+
+        Basically, this just canonicalizes the exponent in a way that is
+        insensitive to global phase. Only relative phases affect the "true"
+        exponent period, and since we omit global phase detail in diagrams this
+        is the appropriate canonicalization to use.
+
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/eigen_gate.py]: Done.

File: cirq/ops/eigen_gate.py [hunk 3]
```diff
@@ -105,6 +105,63 @@ def _with_exponent(self: TSelf,
...
+            A rounded canonicalized exponent.
+        """
+        if not isinstance(self._exponent, (int, float)):
+            return self._exponent
+        result = float(self._exponent)
+
+        # Compute global-phase-independent period of the gate.
+        shifts = list(self._eigen_shifts())
+        relative_shifts = {e - shifts[0] for e in shifts[1:]}
+        relative_periods = [abs(2/e) for e in relative_shifts if e != 0]
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/eigen_gate.py]: Notice that `relative_shifts` was a set, so we're guaranteed that the items are unique by the time we get here.

File: cirq/ops/common_gates_test.py [hunk 4]
```diff
@@ -701,3 +701,40 @@ def test_phase_by_xy():
...
+
+def test_rxyz_circuit_diagram():
+    q = cirq.NamedQubit('q')
+    cirq.testing.assert_has_diagram(
+        cirq.Circuit.from_ops(
+            cirq.Rx(np.pi).on(q),
+            cirq.Rx(-np.pi).on(q),
+            cirq.Rx(-np.pi + 0.00001).on(q),
+            cirq.Rx(-np.pi - 0.00001).on(q),
+            cirq.Rx(3*np.pi).on(q),
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/common_gates_test.py]: It produces `X^0.0`, which is the behavior I intended.

**Pull request merged**